cpu/cc26xx_cc13xx/power_arch.c

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (C) 2020 Locha Inc
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_cc26xx_cc13xx
+ * @{
+ *
+ * @file
+ * @brief       Power management abstractions
+ *
+ * @author      Jean Pierre Dudey <jeandudey@hotmail.com>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+
+#define DOMAIN_ON (1)
+
+/* Save changes of the PRCM */
+static void prcm_commit(void)
+{
+    /* Write CLKLOADCTL in the non-buffered register bank to avoid buffered
+     * writes */
+    PRCM_NONBUF->CLKLOADCTL = CLKLOADCTL_LOAD;
+
+    /* Wait while load is done */
+    while (!(PRCM->CLKLOADCTL & CLKLOADCTL_LOADDONE)) {}
+}
+
+bool power_is_domain_enabled(const power_domain_t domain)
+{
+    switch (domain) {
+        case POWER_DOMAIN_PERIPHERALS:
+            return PRCM->PDSTAT0PERIPH == DOMAIN_ON;
+
+        case POWER_DOMAIN_SERIAL:
+            return PRCM->PDSTAT0SERIAL == DOMAIN_ON;
+
+        case POWER_DOMAIN_RFC:
+            return (PRCM->PDSTAT1RFC == DOMAIN_ON) &&
+                   (PRCM->PDSTAT0RFC == DOMAIN_ON);
+
+        case POWER_DOMAIN_VIMS:
+            return PRCM->PDSTAT1VIMS == DOMAIN_ON;
+
+        case POWER_DOMAIN_CPU:
+            return PRCM->PDSTAT1CPU == DOMAIN_ON;
+
+        default:
+            return false;
+    }
+
+    return false;
+}
+
+void power_enable_domain(const power_domain_t domain)
+{
+    switch (domain) {
+        case POWER_DOMAIN_PERIPHERALS:
+            PRCM->PDCTL0PERIPH = DOMAIN_ON;
+            break;
+
+        case POWER_DOMAIN_SERIAL:
+            PRCM->PDCTL0SERIAL = DOMAIN_ON;
+            break;
+
+        case POWER_DOMAIN_RFC:
+            PRCM->PDCTL0RFC = DOMAIN_ON;
+            PRCM->PDCTL1RFC = DOMAIN_ON;
```

Comment:
Why are there two of them?

Reply:
It seems that on `cc26x0` it's needed to write to both registers, on `cc26x2_cc13x2` it's only necessary to write to `PRCM->PDCTL0RFC` to enable the power domain.

On the SDK for `cc26x0` a comment in [`PRCMPowerDomainOn()`](https://github.com/contiki-ng/coresdk_cc13xx_cc26xx/blob/0d78d3280357416a5c0388148cda13717c9ffaa5/source/ti/devices/cc26x0/driverlib/prcm.c#L433-L437) explains it a little bit:

```
        HWREG(PRCM_BASE + PRCM_O_PDCTL0RFC   ) = 1;
        // The PDCTL1RFC access is meant to "be used by RFC in autonomous mode", but keeping
        // it for compatibility on already ROM'ed products (since this is a ROM function).
        // RFC power domain is on if ( PRCM_O_PDCTL0RFC || PRCM_O_PDCTL1RFC ).
        HWREG(PRCM_BASE + PRCM_O_PDCTL1RFC   ) = 1;
```

That's not present for `cc26x2_cc13x2` SDK. I suppose I can it put behind a `#ifdef` to write it when necessary.

Reply:
So does that mean it's not necessary? Or only necessary on some hardware and it will simply do nothing when not needed?
In that case just keep it, but add the comment.

Reply:
I've tested it and on a cc1312 it just works normally (setting both), only explicitly needed for cc26x0. I'll add a comment.